<a href="https://colab.research.google.com/github/Aashi1102/FlyrankAI-Internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Aashi1102/FlyrankAI-Internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*


**Lane:** Refresh / Content Opportunity Scoring

I chose this lane because it focuses on identifying pages that should be reviewed first based on search and content performance. Instead of manually checking thousands of pages, a data-driven ranking can help prioritize pages that are more likely to benefit from review. This supports better decision-making while saving time and effort.

In [10]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
## 1. My lane (or freestyle) and why
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
else:
    # find the repo root from wherever this kernel started
    while not os.path.isdir("data/raw") and os.getcwd() != "/":
        os.chdir("..")

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "starter CSV not found — are you at the repo root?"
print("Starter data found. You're ready.")

import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Dataset loaded successfully.")
print("Total rows:", len(df))
print("Total columns:", len(df.columns))


Working dir: /content/flyrank-ml-internship-starter/flyrank-ml-internship-starter
Starter data found. You're ready.
Dataset loaded successfully.
Total rows: 30000
Total columns: 44


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

**Research Question**

Which pages should be reviewed first for content refresh based on their search and engagement signals?

**Decision**

Identify which pages deserve priority for review.

**Action**

The SEO or content team can review and update the highest-priority pages first.

**Cost of a Wrong Recommendation**

If the recommendation is incorrect, the team may spend time reviewing pages with little improvement potential or overlook pages that actually need attention.

**Why Data and ML Help**

Large websites may contain thousands of pages. Machine learning can identify useful patterns in historical data and help rank pages for review more efficiently than manual analysis.

In [11]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("Total pages:", len(df))

print("Average impressions:",
      round(df["impressions_90d"].mean(),2))

print("Pages with declining trend:",
      (df["trend_direction"]=="down").sum())

Total pages: 30000
Average impressions: 5200.37
Pages with declining trend: 16262


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

The starter dataset contains thousands of content pages with search performance metrics such as impressions, clicks, CTR, and trend direction.

Out of 30,000 total pages, 16,262 (54.2%) are experiencing a declining trend, with average impressions sitting at 5,200.37. Focusing on ranking these declining pages will help prioritize content updates effectively.


In [12]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

print(df["trend_direction"].value_counts())

print()

print(df["content_type"].value_counts())

trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64

content_type
keyword article       27207
feedly article         2096
comparison article      697
Name: count, dtype: int64


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*



This project is intended to identify patterns in historical search and content data that may help prioritize content review. The results are observational and decision-support only.

This work cannot prove Google's ranking algorithm or guarantee that updating a page will improve its performance. Human review is still required before taking action.

In [13]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("Notebook completed successfully.")

Notebook completed successfully.


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.